pkg/gpu/cuda/fatbin.go

```diff
@@ -0,0 +1,475 @@
+
+	return cp
+}
+
+func (cp *cubinParser) getKernel(name string) *CubinKernel {
+	if _, ok := cp.kernels[name]; !ok {
+		cp.kernels[name] = &CubinKernel{
+			Name: name,
+		}
+	}
+	return cp.kernels[name]
+}
+
+func (cp *cubinParser) parseCubinElf(data []byte) error {
+	// Hacks to be able to parse the ELF: the ELF version is not supported by the Go ELF parser, so we need to
+	// trick it into thinking it's the old version. Check for boundaries first
+	if len(data) < 21 {
+		return fmt.Errorf("invalid cubin data, too short")
+	}
+	data[20] = 1
+
+	cubinElf, err := elf.NewFile(bytes.NewReader(data))
+	if err != nil {
+		return fmt.Errorf("failed to parse cubin ELF: %w", err)
+	}
+	defer cubinElf.Close()
+
+	// Iterate through all the sections, parse all the ones we know how to parse
+	for _, sect := range cubinElf.Sections {
+		for prefix, parser := range cp.sectPrefixToParser {
+			prefixWithDot := prefix + "."
+
+			if !strings.HasPrefix(sect.Name, prefix) {
+				continue
+			}
+
+			var buffer *bytes.Reader
+			if sect.Type != elf.SHT_NOBITS {
+				// do not provide a buffer for NOBITS sections
+				sectData, err := sect.Data()
+				if err != nil {
+					return fmt.Errorf("failed to read section %s: %w", sect.Name, err)
+				}
+				buffer = bytes.NewReader(sectData)
+			}
+
+			var kernelName string
+			if strings.HasPrefix(sect.Name, prefixWithDot) {
+				kernelName = strings.TrimPrefix(sect.Name, prefixWithDot)
+			}
+
+			err = parser(sect, buffer, kernelName)
+			if err != nil {
+				return fmt.Errorf("failed to parse section %s: %w", sect.Name, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// nvInfoParsedItem is the parsed representation of an item in the .nv.info section, including the value
+type nvInfoParsedItem struct {
+	nvInfoItem
+	value []byte
+}
+
+func (cp *cubinParser) parseNvInfoSection(_ *elf.Section, buffer *bytes.Reader, kernelName string) error {
+	items := make(map[nvInfoAttr]nvInfoParsedItem)
+
+	for buffer.Len() > 0 {
+		var item nvInfoItem
+		if err := binary.Read(buffer, binary.LittleEndian, &item); err != nil {
+			return fmt.Errorf("failed to read item: %w", err)
+		}
+
+		var valueSize int
+
+		switch item.Format {
+		case nviFmtBval:
+			valueSize = 1
+		case nviFmtHval:
+			valueSize = 2
+		case nviFmtSval:
+			var nval uint16
+			if err := binary.Read(buffer, binary.LittleEndian, &nval); err != nil {
+				return fmt.Errorf("failed to parse value size: %w", err)
+			}
+			valueSize = int(nval)
+		}
+
+		parsedItem := nvInfoParsedItem{
+			nvInfoItem: item,
+			value:      make([]byte, valueSize),
+		}
+		if _, err := io.ReadFull(buffer, parsedItem.value); err != nil {
+			return fmt.Errorf("failed to read value: %w", err)
+		}
+		items[item.Attr] = parsedItem
+	}
+
+	if kernelName != "" {
+		cp.getKernel(kernelName).attributes = items
+	}
+
+	return nil
+}
+
+func (cp *cubinParser) parseTextSection(sect *elf.Section, _ *bytes.Reader, kernelName string) error {
```

Comment:
It's so that it matches the signature required by the `sectPrefixToParser`, that way we can match each section to the corresponding parser without a chain of `if/elseif/...`

Reply:
suggestion to simplify this:

introduce readSection helper to return Reader over section data:

```
func readSection(sect *elf.Section) (*bytes.Reader, error) {
	var buffer *bytes.Reader
	// do not provide a buffer for NOBITS sections, as they have no data
	if sect.Type != elf.SHT_NOBITS {
		sectData, err := sect.Data()
		if err != nil {
			return nil, fmt.Errorf("failed to read section %s: %w", sect.Name, err)
		}
		buffer = bytes.NewReader(sectData)
	}
	return buffer, nil
}
```

replace the map of section prefixes to callbacks with a switch case:
(now all helper functions can have a single interface of `func(*elf.Section) error`)

		switch {
		case strings.HasPrefix(sect.Name, ".nv.info"):
			cp.parseNvInfoSection(s)
		case strings.HasPrefix(sect.Name, ".text"):
			cp.parseTextSection(s)
		case strings.HasPrefix(sect.Name, ".nv.shared"):
			cp.parseSharedMemSection(s)
		case strings.HasPrefix(sect.Name, ".nv.constant"):
			cp.parseConstantMemSection(s)
		}

call the readSection inside each one of the helpers